FAILED: akka.http.impl.engine.client.HostConnectionPoolSpec.NewPoolImplementation on AkkaHttpEngineNoNetwork should fail a request if the connection stream fails while waiting for request entity bytes

https://jenkins.akka.io:8498/job/akka-http-nightly-against-akka-master/407/testReport/akka.http.impl.engine.client/HostConnectionPoolSpec/NewPoolImplementation_on_AkkaHttpEngineNoNetwork_should_fail_a_request_if_the_connection_stream_fails_while_waiting_for_request_entity_bytes/
  
```
Error Message

Expected OnNext(_), yet no element signaled during 3 seconds

Stacktrace

java.lang.AssertionError: Expected OnNext(_), yet no element signaled during 3 seconds
	at akka.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:325)
	at akka.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:316)
	at akka.stream.testkit.TestSubscriber$Probe.requestNext(StreamTestKit.scala:753)
	at akka.http.impl.engine.client.HostConnectionPoolSpec$$anonfun$testSet$1$TestSetup$1.expectResponseError(HostConnectionPoolSpec.scala:399)
	at akka.http.impl.engine.client.HostConnectionPoolSpec$$anonfun$testSet$1$$anonfun$apply$mcV$sp$11$$anon$7.<init>(HostConnectionPoolSpec.scala:207)
	at akka.http.impl.engine.client.HostConnectionPoolSpec$$anonfun$testSet$1$$anonfun$apply$mcV$sp$11.apply(HostConnectionPoolSpec.scala:193)
	at akka.http.impl.engine.client.HostConnectionPoolSpec$$anonfun$testSet$1$$anonfun$apply$mcV$sp$11.apply(HostConnectionPoolSpec.scala:193)
```

The one above has this log line which might point to the problem:

```
[ERROR] [01/15/2018 09:15:01.744] [HostConnectionPoolSpec-akka.actor.default-dispatcher-4] [akka.actor.ActorSystemImpl(HostConnectionPoolSpec)] [0 (WaitingForResponseDispatch)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [request entity completed] when in state [WaitingForResponseDispatch]
```

There's another error message

```
[ERROR] [01/15/2018 02:25:29.495] [HostConnectionPoolSpec-akka.actor.default-dispatcher-7] [akka.actor.ActorSystemImpl(HostConnectionPoolSpec)] [0 (Idle)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [responseDispatched] when in state [Idle]
	at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:84)
	at akka.http.impl.engine.client.pool.SlotState.onResponseDispatchable(SlotState.scala:64)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anonfun$onResponseDispatchable$1.apply(NewHostConnectionPool.scala:157)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anonfun$onResponseDispatchable$1.apply(NewHostConnectionPool.scala:157)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:175)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:244)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:253)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onResponseDispatchable(NewHostConnectionPool.scala:157)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.onPull(NewHostConnectionPool.scala:86)
```

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
@@ -621,7 +621,7 @@ class HostConnectionPoolSpec extends AkkaSpec(
   /** Transport that uses actual top-level Http APIs to establish a plaintext HTTP connection */
   case object AkkaHttpEngineTCP extends TopLevelApiClientServerImplementation {
     protected override def bindServerSource = Http().bind("localhost", 0)
-    protected def clientConnectionFlow(connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]] =
+    protected def clientConnectionFlow(serverBinding: ServerBinding, connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]] =
       Http().outgoingConnectionUsingTransport(host = "localhost", port = serverBinding.localAddress.getPort, connectionContext = ConnectionContext.noEncryption(), transport = new KillSwitchedClientTransport(connectionKillSwitch))
   }
 
@@ -632,33 +632,32 @@ class HostConnectionPoolSpec extends AkkaSpec(
    */
   case object AkkaHttpEngineTLS extends TopLevelApiClientServerImplementation {
     protected override def bindServerSource = Http().bind("akka.example.org", 0, connectionContext = ExampleHttpContexts.exampleServerContext)
-    protected def clientConnectionFlow(connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]] =
+    protected def clientConnectionFlow(serverBinding: ServerBinding, connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]] =
       Http().outgoingConnectionUsingTransport(host = "akka.example.org", port = serverBinding.localAddress.getPort, connectionContext = ExampleHttpContexts.exampleClientContext, transport = new KillSwitchedClientTransport(connectionKillSwitch))
   }
   abstract class TopLevelApiClientServerImplementation extends ClientServerImplementation {
     def failsHandlerInputWhenHandlerOutputFails: Boolean = false
 
     protected def bindServerSource: Source[Http.IncomingConnection, Future[ServerBinding]]
-    protected def clientConnectionFlow(connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]]
+    protected def clientConnectionFlow(serverBinding: ServerBinding, connectionKillSwitch: SharedKillSwitch): Flow[HttpRequest, HttpResponse, Future[Http.OutgoingConnection]]
 
-    val connectionProbe = TestProbe()
+    override def get(connectionKillSwitch: SharedKillSwitch): BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, Future[Http.OutgoingConnection]] = {
+      val connectionProbe = TestProbe()
+      system.log.debug("Binding server for test ...")
+      val serverBinding: ServerBinding =
+        bindServerSource
+          .to(Sink.foreach { serverConnection ⇒
+            connectionProbe.ref ! serverConnection
+          })
+          .run().awaitResult(3.seconds)
+      system.log.debug(s"Server bound to [${serverBinding.localAddress}]")
 
-    system.log.debug("Binding server for test ...")
-    val serverBinding: ServerBinding =
-      bindServerSource
-        .to(Sink.foreach { serverConnection ⇒
-          connectionProbe.ref ! serverConnection
-        })
-        .run().awaitResult(3.seconds)
-    system.log.debug(s"Server bound to [${serverBinding.localAddress}]")
-
-    override def get(connectionKillSwitch: SharedKillSwitch): BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, Future[Http.OutgoingConnection]] =
       // needs to be an involved two step process:
       //   1. setup client flow and proxies on the server side to be able to return that flow immediately
       //   2. when client connection was established, grab server connection as well and attach to proxies
       //      (cannot be implemented with just mapMaterializedValue because there's no transposing constructor for BidiFlow)
       BidiFlow.fromGraph(
-        GraphDSL.create(Sink.asPublisher[HttpResponse](fanout = false), Source.asSubscriber[HttpRequest], clientConnectionFlow(connectionKillSwitch))((_, _, _)) { implicit builder ⇒ (resIn, reqOut, client) ⇒
+        GraphDSL.create(Sink.asPublisher[HttpResponse](fanout = false), Source.asSubscriber[HttpRequest], clientConnectionFlow(serverBinding, connectionKillSwitch))((_, _, _)) { implicit builder ⇒ (resIn, reqOut, client) ⇒
           import GraphDSL.Implicits._
 
           builder.materializedValue ~> Sink.foreach[(Publisher[HttpResponse], Subscriber[HttpRequest], Future[Http.OutgoingConnection])] {
@@ -674,6 +673,7 @@ class HostConnectionPoolSpec extends AkkaSpec(
           BidiShape(resIn.in, client.out, client.in, reqOut.out)
         }
       ).mapMaterializedValue(_._3)
+    }
   }
 
   /** Generates a new unique outgoingConnection */

